### ISSUE TITLE
Missing packet

### ISSUE THREAD

**Issue by ghost** [google/perfetto#33]
While working on a [GPU data source](https://gitlab.freedesktop.org/Fahien/gfx-pps/), I noticed that one packet was missing. To be precise, the missing packet is the one before the last packet emitted within the `OnStop` callback.

To demonstrate that, I pushed a [commit on my fork](https://github.com/antonio-caggiano-collabora/perfetto/commit/7721e70ed9bc4a82318cb9586e8ace5a3d53da2c). This commit modifies slightly `test/client_api_example.cc` as described:
- proper packet timestamps
- mark last packet with a test string
- set duration to 4 seconds in `client_api.cfg`

With this modifications, I run a tracing session:

- This is the output of the client. From the logs I would expect to see `4 + 1` (while stopping) packets
  ```j
  [190.534] lient_api_example.cc:66 OnStart called
  [190.660] ient_api_example.cc:107 Tracing lambda called
  [191.661] ient_api_example.cc:107 Tracing lambda called
  [192.661] ient_api_example.cc:107 Tracing lambda called
  [193.662] ient_api_example.cc:107 Tracing lambda called
  [194.535] lient_api_example.cc:69 OnStop called
  [194.535] lient_api_example.cc:76 Tracing lambda called while stopping
  ```
 - This is the trace in text format ([trace.pbtext.txt](https://github.com/google/perfetto/files/4731560/trace.pbtext.txt)). Here I am highlighting the relevant part, where you could notice `3 + 1` (while stopping) packet. One packet is missing, ideally with timestamp `~21193662000000`.
```java
packet {
  trusted_uid: 1000
  timestamp: 21190660987774
  trusted_packet_sequence_id: 2
  previous_packet_dropped: true
  gpu_counter_event {
    counters {
      counter_id: 1
    }
  }
}
packet {
  trusted_uid: 1000
  timestamp: 21191661333029
  trusted_packet_sequence_id: 2
  gpu_counter_event {
    counters {
      counter_id: 1
    }
  }
}
packet {
  trusted_uid: 1000
  timestamp: 21192661807648
  trusted_packet_sequence_id: 2
  gpu_counter_event {
    counters {
      counter_id: 1
    }
  }
}
packet {
  trusted_uid: 1000
  timestamp: 21194535913903
  trusted_packet_sequence_id: 3
  previous_packet_dropped: true
  for_testing {
    str: "Last packet"
  }
}
```

**Comment by betasheet** [comment]
The OnStop() of the data source will be called on a different thread, which uses a different TraceWriter and thus a different packet sequence.

Currently, the only way to guarantee that the last packet from the main thread is picked up would be to tell the main thread (or any others, depending on your setup) to Flush() when OnStop() is called via some cross-thread signal, wait for that to happen, and then call stop_closure() asynchronously.

If you don't call Flush() on each of your instrumented threads, the tracing service falls back to scraping the shared memory buffer chunks for each of these threads. When doing this, it currently has no way of knowing whether the last packet was fully written by the respective thread, so ignores it.

**Comment by primiano** [comment]
`test/client_api_example.cc` has a Flush() on stop though,soon after commit, how come that doesn't work?

**Comment by betasheet** [comment]
Again, that Flush executes in OnStop() on a different thread - it won't affect the main thread's TraceWriter.

**Comment by primiano** [comment]
But here is seems that the packet that gets ` previous_packet_dropped: true` is the one written on OnStop(), where we called Flush()
What am I missing?

**Comment by betasheet** [comment]
So AFAIU, this bug is about the fourth packet on the main thread going missing. See above for the explanation why.

The "Last packet" in the trace (which is emitted during OnStop) is also the first packet in the packet sequence with ID 3. As such, the service marks it with "previous_packet_dropped" (b/c the service is pessimistically saying "there could have been more before this packet on this sequence, I can't tell for sure" - e.g. because the buffer may have wrapped).

We could improve the service's logic for setting this flag (e.g. detect if any chunk of specific sequence was overridden in the buffer due to wrapping, or take into account the buffering mode / whether any wrapping occurred at all to cover some of the more basic cases where there can't have been buffer data loss). Right now, the service simply marks the first packet on each sequence this way regardless of other service state.

**Comment by primiano** [comment]
Ahh okay let's summarize all this now makes more sense:

1. The missing packet is the last written by the main thread, NOT the one written OnStop(), because of lack of Flush()
2. The packet written by OnStop() it makes into the trace but is marked as `previous_packet_dropped: true` because of that long standing bug. 
Q: does this have other side effects? will this be processed anyways?

**Comment by betasheet** [comment]
Reg 2, no side effects (other than clearing incremental state, which is a no-op for the first packet), and yes it will be processed regularly.

**Comment by ddiproietto** [comment]
This has been fixed now in v43.1

**Comment by LalitMaganti** [comment]
🎉🎉🎉🎉🎉
